File: src/codegen/Io.py
"""
I/O symbol definitions for built-in functions.
"""

from utils.AST import IntType, FloatType, StringType, VoidType, BoolType
from codegen.Utils import FunctionType, Symbol, CName


LIB_NAME = "io"

# Built-in functions:
IO_SYMBOL_LIST = [
    # MiniRust specification functions
    Symbol("print_i32", FunctionType([IntType()], VoidType()), CName(LIB_NAME)),
    Symbol("println_i32", FunctionType([IntType()], VoidType()), CName(LIB_NAME)),
    Symbol("print_f32", FunctionType([FloatType()], VoidType()), CName(LIB_NAME)),
    Symbol("println_f32", FunctionType([FloatType()], VoidType()), CName(LIB_NAME)),
    Symbol("print_bool", FunctionType([BoolType()], VoidType()), CName(LIB_NAME)),
    Symbol("println_bool", FunctionType([BoolType()], VoidType()), CName(LIB_NAME)),
    Symbol("print_string", FunctionType([StringType()], VoidType()), CName(LIB_NAME)),
    Symbol("println_string", FunctionType([StringType()], VoidType()), CName(LIB_NAME)),
    Symbol("println", FunctionType([], VoidType()), CName(LIB_NAME)),
]

File: src/runtime/io.java
/**
 * IO class for MiniRust runtime.
 * Provides input/output operations for MiniRust programs.
 */
public class io {
    // MiniRust specification functions
    public static void print_i32(int value) {
        System.out.print(value);
    }
    public static void println_i32(int value) {
        System.out.println(value);
    }
    public static void print_f32(float value) {
        System.out.print(value);
    }
    public static void println_f32(float value) {
        System.out.println(value);
    }
    public static void print_bool(boolean value) {
        System.out.print(value);
    }
    public static void println_bool(boolean value) {
        System.out.println(value);
    }
    public static void print_string(String value) {
        System.out.print(value);
    }
    public static void println_string(String value) {
        System.out.println(value);
    }
    public static void println() {
        System.out.println();
    }
}
